src/hotspot/share/runtime/thread.hpp

```diff
@@ -631,6 +631,31 @@ class Thread: public ThreadShadow {
     assert(_wx_state == expected, "wrong state");
   }
 #endif // __APPLE__ && AARCH64
+
+ private:
+  bool _in_asgct = false;
+ public:
+  bool in_asgct() const { return _in_asgct; }
+  void set_in_asgct(bool value) { _in_asgct = value; }
+  static bool current_in_asgct() {
+    Thread *cur = Thread::current();
```

Comment:
You need to use`current_or_null_safe` here as you may be in a signal handling context.